src/pool/srv_target.c

```diff
@@ -1830,6 +1830,117 @@ update_child_map(void *data)
 	return 0;
 }
 
+static int
+map_bc_create(crt_context_t ctx, struct pool_map *map, struct ds_pool_map_bc **map_bc_out)
+{
+	struct ds_pool_map_bc *map_bc;
+	d_iov_t                map_iov;
+	d_sg_list_t            map_sgl;
+	int                    rc;
+
+	D_ALLOC_PTR(map_bc);
+	if (map_bc == NULL) {
+		rc = -DER_NOMEM;
+		goto err;
+	}
+
+	map_bc->pmc_ref = 1;
+
+	rc = pool_buf_extract(map, &map_bc->pmc_buf);
+	if (rc != 0) {
+		DL_ERROR(rc, "failed to extract pool map buffer");
+		goto err_map_bc;
+	}
+
+	d_iov_set(&map_iov, map_bc->pmc_buf, pool_buf_size(map_bc->pmc_buf->pb_nr));
+	map_sgl.sg_nr     = 1;
+	map_sgl.sg_nr_out = 0;
+	map_sgl.sg_iovs   = &map_iov;
+
+	rc = crt_bulk_create(ctx, &map_sgl, CRT_BULK_RO, &map_bc->pmc_bulk);
+	if (rc != 0)
+		goto err_buf;
+
+	*map_bc_out = map_bc;
+	return 0;
+
+err_buf:
+	D_FREE(map_bc->pmc_buf);
+err_map_bc:
+	D_FREE(map_bc);
+err:
+	return rc;
+}
+
+static void
+map_bc_get(struct ds_pool_map_bc *map_bc)
+{
+	map_bc->pmc_ref++;
+}
+
+static void
+map_bc_put(struct ds_pool_map_bc *map_bc)
+{
+	map_bc->pmc_ref--;
+	if (map_bc->pmc_ref == 0) {
+		crt_bulk_free(map_bc->pmc_bulk);
+		D_FREE(map_bc->pmc_buf);
+		D_FREE(map_bc);
+	}
+}
+
+int
+ds_pool_lookup_map_bc(struct ds_pool *pool, crt_context_t ctx, struct ds_pool_map_bc **map_bc_out,
+		      uint32_t *map_version_out)
+{
+	struct ds_pool_map_bc *map_bc;
+	uint32_t               map_version;
+
+	D_ASSERT(dss_get_module_info()->dmi_xs_id == 0);
+
+	/* For accessing pool->sp_map, but not really necessary. */
+	ABT_rwlock_rdlock(pool->sp_lock);
+
+	if (pool->sp_map == NULL) {
+		ABT_rwlock_unlock(pool->sp_lock);
+		return -DER_NONEXIST;
+	}
+
+	if (pool->sp_map_bc == NULL) {
+		int rc;
+
+		rc = map_bc_create(ctx, pool->sp_map, &pool->sp_map_bc);
+		if (rc != 0) {
+			ABT_rwlock_unlock(pool->sp_lock);
+			return rc;
+		}
+	}
+
+	map_bc_get(pool->sp_map_bc);
+	map_bc = pool->sp_map_bc;
+	map_version = pool_map_get_version(pool->sp_map);
+
+	ABT_rwlock_unlock(pool->sp_lock);
+
+	*map_bc_out = map_bc;
+	*map_version_out = map_version;
+	return 0;
+}
+
+void
+ds_pool_put_map_bc(struct ds_pool *pool, struct ds_pool_map_bc *map_bc)
+{
+	D_ASSERT(dss_get_module_info()->dmi_xs_id == 0);
+
+	/* We could cache this longer, actually. */
```

Comment:
yeah, I don't quite see why we have to invalidate the bulk cache here? I think we just need to invalid it when changing pool map, no? (the change you made oin ds_pool_tgt_map_update()).

Reply:
also not quite understand why map_bc_put() 2 times below if (pool->sp_map_bc == map_bc)? then the cache is invalidated even no pool map change right?

Reply:
Yes, there's problem. Let me update...

Reply:
the logic looks correct although.